### PR TITLE
feat: enable next button

### PR DESCRIPTION
**Pull request by BhuvaneshPatil**
### Details

Currently we don't have next button for distance request enabled by default, this PR will enable the distance request and display the error in case there is an error.

### Fixed Issues

$ https://github.com/Expensify/App/issues/27045
PROPOSAL: https://github.com/Expensify/App/issues/27045#issuecomment-1711984411

#### Test 1
1. Empty waypoints
2. Click next
3. Verify an error appears asking you to enter waypoints

#### Test 2
1. Start waypoint A
2. empty waypoint
3. Finish waypoint B
4. Click next
5. Verify that you are navigated without errors

#### Test 3
1. start waypoint A
2. finish waypoint A
3. Click next
4. Verify an error appears about duplicate waypoints

#### Test 4
1. start waypoint A
2. stop waypoint A
3. finish waypoint B
4. Click next
5. Verify that you are navigated without errors

#### Test 5
1. start waypoint A
2. stop waypoint B
3. stop waypoint B
4. finish waypoint A
5. Click next
6. Verify that you are navigated without error

- [X] Verify that no errors appear in the JS console

### QA Steps

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C01GTK53T8Q/p1696910414908269
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/27822551/187beb97-8ccc-4eaf-beb7-75c22fc54294

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/27822551/58594af1-853f-4970-a563-988976219a54

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/27822551/93155fdb-0601-4167-be1b-bc85333a6da8

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/27822551/d17e2190-8c85-4004-9eac-091490bb89b3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/27822551/410d5126-1572-40b7-ac6d-f4039c15615a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/27822551/2dac22f9-8001-44ea-938a-aa942e956800

</details>

### CODE REVIEW

**Comment by shawnborton** [comment]
Looking good, thanks!

**Review by allroundexperts — APPROVED** [review]
Looks good!

**Comment by shawnborton** [comment]
Yup, lgtm too

File: src/components/DistanceRequest/index.js [hunk 1]
```diff
@@ -101,6 +104,9 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
 
         // Create the initial start and stop waypoints
         Transaction.createInitialWaypoints(transactionID);
+        return () => {
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/components/DistanceRequest/index.js:107]: NAB: Thanks for commenting on the reason you added this. Next time, please add that as a code comment to give context into _why_ that logic is here.

File: src/components/DistanceRequest/index.js [hunk 2]
```diff
@@ -118,6 +124,13 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         scrollViewRef.current.scrollToEnd({animated: true});
     }, [numberOfPreviousWaypoints, numberOfWaypoints]);
 
+    useEffect(() => {
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/components/DistanceRequest/index.js:127]: OK, here is another one. At this point, I'm going to request that you add these contextual comments to the code.

File: src/components/DistanceRequest/index.js [hunk 3]
```diff
@@ -130,6 +143,21 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         Navigation.navigate(isEditingRequest ? ROUTES.MONEY_REQUEST_EDIT_WAYPOINT.getRoute(report.reportID, transactionID, index) : ROUTES.MONEY_REQUEST_WAYPOINT.getRoute('request', index));
     };
 
+    const getError = () => {
+        // get route error if available else show the invalid number of waypoints error
```
Comment (tgolen) [inline src/components/DistanceRequest/index.js:147]: Comments should start with a capital letter and end with a period (it should be a normal sentence). Please update this and the comment below as well.

File: src/components/DistanceRequest/index.js [hunk 4]
```diff
@@ -130,6 +143,21 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
...
+        if (hasRouteError) {
+            return ErrorUtils.getLatestErrorField(transaction, 'route');
+        }
+
+        if (_.size(validatedWaypoints) < 2) {
+            // initially, both waypoints will be null, and if we give fallback value as emmpty string that will result in true condition, that's why different default values
```
Comment (tgolen) [inline src/components/DistanceRequest/index.js:153]:
```suggestion
            // initially, both waypoints will be null, and if we give fallback value as empty string that will result in true condition, that's why different default values
```
typo

File: src/components/DistanceRequest/index.js [hunk 5]
```diff
@@ -130,6 +143,21 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         Navigation.navigate(isEditingRequest ? ROUTES.MONEY_REQUEST_EDIT_WAYPOINT.getRoute(report.reportID, transactionID, index) : ROUTES.MONEY_REQUEST_WAYPOINT.getRoute('request', index));
     };
 
+    const getError = () => {
+        // get route error if available else show the invalid number of waypoints error
+        if (hasRouteError) {
+            return ErrorUtils.getLatestErrorField(transaction, 'route');
+        }
+
+        if (_.size(validatedWaypoints) < 2) {
```
Comment (tgolen) [inline src/components/DistanceRequest/index.js:152]: It's confusing here that this is mixing `validatedWaypoints` and `waypoints` inside the `if` statement. Can you please clarify why both variables are used for this error?

File: src/components/DistanceRequest/index.js [hunk 6]
```diff
@@ -183,13 +211,27 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
...
                     pressOnEnter
                     style={[styles.w100, styles.mb4, styles.ph4, styles.flexShrink0]}
-                    onPress={() => onSubmit(waypoints)}
-                    isDisabled={_.size(validatedWaypoints) < 2 || (!isOffline && (hasRouteError || isLoadingRoute || isLoading))}
+                    onPress={() => {
+                        setUserTriedSubmitting(true);
+                        // if there is any error or loading state, don't let user go to next page
+                        if (_.size(validatedWaypoints) < 2 || hasRouteError || isLoadingRoute || isLoading) {
+                            return;
+                        }
```
Comment (tgolen) [inline src/components/DistanceRequest/index.js:232]: Could you please move this logic inside of `onSubmit`?

File: src/languages/en.ts [hunk 7]
```diff
@@ -567,6 +567,8 @@ export default {
             genericDeleteFailureMessage: 'Unexpected error deleting the money request, please try again later',
             genericEditFailureMessage: 'Unexpected error editing the money request, please try again later',
             genericSmartscanFailureMessage: 'Transaction is missing fields',
+            duplicateWaypointsErrorMessage: 'There are duplicate waypoints',
```
Comment (tgolen) [inline src/languages/en.ts:570]: Shouldn't this message be more like "Please remove duplicate waypoints"?

File: src/components/DistanceRequest/index.js [hunk 8]
```diff
@@ -68,6 +70,7 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
     const {translate} = useLocalize();
 
     const [optimisticWaypoints, setOptimisticWaypoints] = useState(null);
+    const [userTriedSubmitting, setUserTriedSubmitting] = useState(false);
```
[the remark this answers is not in the document]
Comment (tgolen) [reply src/components/DistanceRequest/index.js:73]: I agree with your first instinct @allroundexperts and I don't like this being tied to a user action. I'd rather prefer that it is tied to an error state. Maybe something like `hasError, setHasError`.

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/components/DistanceRequest/index.js (hunk 6 above)
Comment (tgolen) [reply src/components/DistanceRequest/index.js]: Ah, I didn't realize that `onSubmit()` was from props. OK, then the proper thing to do is to move this to a function that is defined above the `return`, add a `useCallback()` and pass the function here as a reference.

File: src/components/DistanceRequest/index.js (hunk 5 above)
Comment (tgolen) [reply src/components/DistanceRequest/index.js:154]: OK, but why check `validatedWaypoints` at all then?

Comment (tgolen) [reply src/components/DistanceRequest/index.js:154]: Ah, yes that makes more sense!

File: src/components/DistanceRequest/index.js [hunk 9]
```diff
@@ -151,6 +182,15 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         [transactionID, waypoints, waypointsList],
     );
 
+    const onPress = useCallback(() => {
```
Comment (tgolen) [inline src/components/DistanceRequest/index.js:185]: Please rename this to what the function does like `submitWaypoints()`

**Review by tgolen — CHANGES_REQUESTED** [review]
Looks good, just this one last thing.

**Review by tgolen — APPROVED** [review]

**Comment by tgolen** [comment]
I was just waiting for the tests to pass. Not sure why it was taking the lint test so long.

File: src/components/DistanceRequest/index.js [hunk 10]
```diff
@@ -151,6 +182,15 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         [transactionID, waypoints, waypointsList],
     );
 
+    const submitWaypoints = useCallback(() => {
+        // If there is any error or loading state, don't let user go to next page.
+        if (_.size(validatedWaypoints) < 2 || hasRouteError || isLoadingRoute || isLoading) {
```
Comment (allroundexperts) [inline src/components/DistanceRequest/index.js:187]: This condition should have taken into account if the user was offline or not. This later caused https://github.com/Expensify/App/issues/32397

File: src/components/DistanceRequest/index.js [hunk 11]
```diff
@@ -130,6 +145,22 @@ function DistanceRequest({transactionID, report, transaction, route, isEditingRe
         Navigation.navigate(isEditingRequest ? ROUTES.MONEY_REQUEST_EDIT_WAYPOINT.getRoute(report.reportID, transactionID, index) : ROUTES.MONEY_REQUEST_WAYPOINT.getRoute('request', index));
     };
 
+    const getError = () => {
```
Comment (alitoshmatov) [inline src/components/DistanceRequest/index.js:148]: Duplicate waypoint is not showing error when rearranged, ref: https://github.com/Expensify/App/pull/35910

**Pull request merged**